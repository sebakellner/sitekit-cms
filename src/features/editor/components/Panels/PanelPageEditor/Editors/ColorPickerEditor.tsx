import {
  Box,
  ColorPicker,
  Field,
  HStack,
  parseColor,
  Portal,
} from '@chakra-ui/react'

import type { PanelEditorBase } from '@features/editor/types/panelEditors.types'

const ColorPickerEditor = ({ title, value, onChange }: PanelEditorBase) => {
  return (
    <Box mb={3}>
      <Field.Root orientation="horizontal">
        <Field.Label fontSize="xs">{title}</Field.Label>
        <ColorPicker.Root
          value={value ? parseColor(value) : undefined}
          format="rgba"
          onValueChange={(e) => onChange(e.valueAsString)}
          w="100%"
        >
          <ColorPicker.HiddenInput />
          <ColorPicker.Control>
            <ColorPicker.Input />
            <ColorPicker.Trigger />
          </ColorPicker.Control>
          <Portal>
            <ColorPicker.Positioner>
              <ColorPicker.Content>
                <ColorPicker.Area />
                <HStack>
                  <ColorPicker.EyeDropper size="xs" variant="outline" />
                  <ColorPicker.Sliders />
                </HStack>
              </ColorPicker.Content>
            </ColorPicker.Positioner>
          </Portal>
        </ColorPicker.Root>
      </Field.Root>
    </Box>
  )
}

export default ColorPickerEditor
