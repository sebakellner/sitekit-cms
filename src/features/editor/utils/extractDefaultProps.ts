import { ComponentMetaSchema } from '@components/site/schemas/componentMeta.schema'
import type { ComponentMeta } from '@components/site/types'

function isDefaultProp(val: unknown): val is { default: unknown } {
  return typeof val === 'object' && val !== null && 'default' in val
}

export function extractDefaultProps(
  props: ComponentMeta['props'] | undefined,
  fallback: unknown = undefined
): Record<string, unknown> {
  if (!props) return {}
  const result = ComponentMetaSchema.shape.props.safeParse(props)
  if (!result.success) {
    console.error('Invalid props schema:', result.error)
    return {}
  }
  const validProps = result.data
  return Object.fromEntries(
    Object.entries(validProps).map(([key, val]) => [
      key,
      isDefaultProp(val) ? val.default : fallback,
    ])
  )
}
