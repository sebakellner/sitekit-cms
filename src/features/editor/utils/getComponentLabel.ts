import React from 'react'

export function getComponentLabel(
  children: React.ReactNode,
  fallback = 'Component'
): string {
  if (React.isValidElement(children)) {
    const type = children.type
    if (typeof type === 'function' && type.name) return type.name
  }
  return fallback
}
