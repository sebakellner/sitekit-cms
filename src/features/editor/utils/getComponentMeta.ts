import { ComponentMetaSchema } from '../schemas/componentMeta.schema'

type ModuleWithDefault = { default: unknown }

export function getComponentMeta(module: unknown) {
  if (
    module &&
    typeof module === 'object' &&
    'default' in module &&
    (module as ModuleWithDefault).default &&
    typeof (module as ModuleWithDefault).default === 'object'
  ) {
    const def = (module as ModuleWithDefault).default
    const result = ComponentMetaSchema.safeParse(def)
    return {
      success: result.success,
      data: result.success ? result.data : null,
      error: result.success ? null : result.error,
    }
  }
  return {
    success: false,
    data: null,
    error: new Error(
      'Invalid module format: Module must have a default export with valid component metadata'
    ),
  }
}
