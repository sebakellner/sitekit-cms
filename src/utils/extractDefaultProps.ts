import type { ComponentMeta } from '@components/site/types'

export function extractDefaultProps(
  props: ComponentMeta['props'] | undefined
): Record<string, unknown> {
  if (!props) return {}
  return Object.fromEntries(
    Object.entries(props).map(([key, val]) => [
      key,
      (val as { default: unknown }).default,
    ])
  )
}
