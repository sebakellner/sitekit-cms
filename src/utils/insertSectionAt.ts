import type { Section } from '@src/types'

export function insertSectionAtUtil(
  sections: Section[],
  setSections: (sections: Section[]) => void,
  id: string,
  index: number
) {
  const sectionId = String(id).startsWith('selector-')
    ? String(id).replace('selector-', '')
    : id
  const uniqueId = `${sectionId}-${Date.now()}`
  const newSection = {
    id: uniqueId,
    name: sectionId,
    props: {},
  }
  const newSections = [...sections]
  newSections.splice(index, 0, newSection)
  setSections(newSections)
}
