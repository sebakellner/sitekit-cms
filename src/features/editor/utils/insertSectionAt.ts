import { componentsRegistry } from '@registry/componentRegistry'
import { v4 as uuidv4 } from 'uuid'

import type { ComponentMeta } from '@components/site/types'
import { extractDefaultProps } from '@src/features/editor/utils/extractDefaultProps'
import type { Section } from '../schemas/section.schema'

export async function insertSectionAtUtil(
  sections: Section[],
  setSections: (sections: Section[]) => void,
  name: string,
  index: number
) {
  const loader = componentsRegistry[name as keyof typeof componentsRegistry]
  if (!loader) {
    console.warn(`Component "${name}" not found in registry.`)
    return
  }

  const metaModule = await loader()

  const meta = metaModule.default as ComponentMeta

  const defaultProps = extractDefaultProps(meta?.props)

  const newSection: Section = {
    id: uuidv4(),
    name,
    props: defaultProps,
  }

  const updatedSections = [
    ...sections.slice(0, index),
    newSection,
    ...sections.slice(index),
  ]

  setSections(updatedSections)
}
