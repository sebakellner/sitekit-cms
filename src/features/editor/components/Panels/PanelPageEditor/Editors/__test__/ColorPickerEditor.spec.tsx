import { describe, test, expect } from 'vitest'
import { render } from 'test/utils/render'
import ColorPickerEditor from '../ColorPickerEditor'
import type { PanelEditorBase } from '@features/editor/types/panelEditors.types'
import { screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { act } from 'react'

const setup = (overrides: Partial<PanelEditorBase<string>> = {}) => {
  const defaultProps: PanelEditorBase<string> = {
    title: 'Color Picker',
    value: '#ffffff',
    onChange: () => {},
  }
  const props = { ...defaultProps, ...overrides }
  return <ColorPickerEditor {...props} />
}

describe('ColorPicker', () => {
  test('should match with snapshot', () => {
    const { container } = render(setup())
    expect(container).toMatchSnapshot()
  })

  test('should handle falsy value and pass undefined to ColorPicker.Root', () => {
    render(setup({ value: undefined }))
    const input = screen.getByLabelText('Color Picker')
    expect(input).toHaveValue('')
  })

  test('should call onChange with valueAsString when color changes', async () => {
    const handleChange = vi.fn()
    render(
      setup({
        onChange: handleChange,
      })
    )

    const input = screen.getByLabelText('hex')
    await userEvent.clear(input)
    await userEvent.type(input, '#ff0000')
    await userEvent.tab()

    await waitFor(() => {
      expect(handleChange).toHaveBeenCalled()
    })
  })
})
