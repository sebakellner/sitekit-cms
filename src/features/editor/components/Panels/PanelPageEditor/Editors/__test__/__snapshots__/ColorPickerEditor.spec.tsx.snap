// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ColorPicker > should match with snapshot 1`] = `
<div>
  <script
    nonce=""
  >
    ((e,i,s,u,m,a,l,h)=&gt;{let d=document.documentElement,w=["light","dark"];function p(n){(Array.isArray(e)?e:[e]).forEach(y=&gt;{let k=y==="class",S=k&&a?m.map(f=&gt;a[f]||f):m;k?(d.classList.remove(...S),d.classList.add(a&&a[n]?a[n]:n)):d.setAttribute(y,n)}),R(n)}function R(n){h&&w.includes(n)&&(d.style.colorScheme=n)}function c(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(u)p(u);else try{let n=localStorage.getItem(i)||s,y=l&&n==="system"?c():n;p(y)}catch(n){}})("class","theme","system",null,["light","dark"],null,true,true)
  </script>
  <div
    class="css-cdgtm1"
  >
    <div
      class="chakra-field__root css-1anu6qw"
      data-part="root"
      data-scope="field"
      id="field::«r0»"
      orientation="horizontal"
      role="group"
    >
      <label
        class="chakra-field__label css-84rnco"
        data-part="label"
        data-scope="field"
        for="«r0»"
        id="field::«r0»::label"
      >
        Color Picker
      </label>
      <div
        class="colorPicker__root css-zxnr2f"
        data-part="root"
        data-scope="color-picker"
        dir="ltr"
        id="color-picker:«r1»"
        style="--value: rgba(255, 255, 255, 1);"
      >
        <input
          id="color-picker:«r1»:hidden-input"
          style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap; word-wrap: normal;"
          tabindex="-1"
          type="text"
          value="rgba(255, 255, 255, 1)"
        />
        <div
          class="colorPicker__control css-1q6y153"
          data-part="control"
          data-scope="color-picker"
          data-state="closed"
          dir="ltr"
          id="color-picker:«r1»:control"
        >
          <input
            aria-label="hex"
            autocomplete="off"
            class="colorPicker__channelInput css-1aemz6e"
            data-channel="hex"
            data-part="channel-input"
            data-scope="color-picker"
            dir="ltr"
            max="16777215"
            min="0"
            spellcheck="false"
            step="1"
            style="appearance: none;"
            type="text"
            value="#FFFFFF"
          />
          <button
            aria-controls="color-picker:«r1»:content"
            aria-label="select color. current color is rgba(255, 255, 255, 1)"
            aria-labelledby="field::«r0»::label"
            class="colorPicker__trigger css-10mv649"
            data-part="trigger"
            data-scope="color-picker"
            data-state="closed"
            dir="ltr"
            id="color-picker:«r1»:trigger"
            style="position: relative;"
            type="button"
          >
            <div
              class="colorPicker__swatch css-lkq982"
              data-part="swatch"
              data-scope="color-picker"
              data-state="checked"
              data-value="#FFFFFF"
              dir="ltr"
              style="--color: rgba(255, 255, 255, 1); position: relative; background: rgb(255, 255, 255);"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
